python 3 compatibility

jtprince opened the pull request:
Updates to make this package compatible with python3.  Also, updated the dependencies to latest stable versions.

All tests passing on both python 2.7 and 3.4.

Assuming you approve of this request, please also release the updated package to pypi.  Thanks!

jtprince committed:
fix most python3 compatibility issues

jtprince committed:
allow git to ignore temporary files

jtprince committed:
fix all but one failing tests for python2/3 compatibility

Deals with various string / bytestring issues while making this python 2/3
compatible.

jtprince committed:
Update the readme to reflect python 3.4 compatibility

jtprince committed:
version bump

jtprince committed:
added future dependency and fixed gitignore

gnarea commented on .gitignore:
```diff
@@ -0,0 +1,3 @@
+__pycache__
```
I think these additions shouldn't be here. They should be in your global `.gitignore`.

jtprince replied:
sounds good

gnarea commented on hubspot/connection/__init__.py:
```diff
@@ -228,11 +230,11 @@ def __exit__(self, exc_type, exc_value, traceback):
 
 
-_AuthenticationKey = Record.create_type('_AuthenticationKey', 'key_value')
+_AuthenticationKey = Record.create_type(str('_AuthenticationKey'), str('key_value'))
```
Why do you need `str()`?

jtprince replied:
Without the str(), and with the additions that make it python 3 compatibility, you end up with this kind of error:

Traceback (most recent call last):
  File "/home/jtprince/virtualenvs/python2.7-playground/local/lib/python2.7/site-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/home/jtprince/virtualenvs/python2.7-playground/local/lib/python2.7/site-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/home/jtprince/virtualenvs/python2.7-playground/local/lib/python2.7/site-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/jtprince/dev/hubspot-connection/tests/test_testing_utils.py", line 24, in <module>
    from hubspot.connection.exc import HubspotAuthenticationError
  File "/home/jtprince/dev/hubspot-connection/hubspot/connection/**init**.py", line 233, in <module>
    _AuthenticationKey = Record.create_type('_AuthenticationKey', 'key_value')
  File "build/bdist.linux-x86_64/egg/pyrecord/__init__.py", line 204, in create_type
    **default_values_by_field_name
  File "build/bdist.linux-x86_64/egg/pyrecord/**init**.py", line 239, in extend_type
    default_values_by_field_name,
  File "build/bdist.linux-x86_64/egg/pyrecord/**init**.py", line 245, in _create_type
    record_type = type(type_name, (cls,), {})
TypeError: type() argument 1 must be string, not unicode

str() solves the issue by giving pyrecord a bytestring instead of unicode under python 2.7 and giving  pyrecord unicode instead of a bytestring in python 3.4.  I believe it is the right way to solve this problem while achieving compatibility (although I'm always open to suggestions, especially when it's your code base ;)

gnarea replied:
That error is very strange. I'm the author of PyRecord, and the test I wrote for that case works on Python 2.7/3.3/3.4/3.5: https://github.com/gnarea/pyrecord/blob/11aa1625b810aa025d3c0edd3f2c04375d5e46b0/tests/test_record_type.py#L25

I'd expect the expression `Record.create_type('AuthenticationKey', 'key_value')` to work on Python 2 and 3. I'll investigate that, but for now I'd suggest to remove these `str()` calls from here as I could deal with that case in PyRecord if necessary.

jtprince replied:
Sounds good.  I'll remove these.  I agree that it is a little baffling and that the best place to deal with it is in pyrecord.

jtprince replied:
I'm not sure why, but this time after removing those I am not getting the same errors.  So, that's fantastic, I think (except not knowing _why_ the behavior changed).

LuRsT replied:
Ok, I found the problem here, by doing `from __future__ import unicode_literals` every string becomes unicode which causes that problem when in `pyrecord` we pass the name into `type()`. I'm currently trying to solve this so we can merge your code, thank you for your patience :smiley:

gnarea commented on hubspot/connection/exc.py:
```diff
@@ -17,7 +17,10 @@
-    pass
+    def __init__(self, message):
+        """ Explicitly makes the message available as the 'message' attribute. """
+        self.message = message
```
I think this is probably not necessary as you have `str(exception)`.

UPDATE: I can now see why you did this. It's the fact that `str()` proxies `repr()` below. We shouldn't have done that -- I don't think we should've overridden `str()`, which would make it possible to use my suggestion above.

jtprince replied:
I see.  I'll play around with removing the overriden str() and also removing my explicit override.

jtprince replied:
I tried a few ways of making this ("removing the overriden str() and also removing my explicit override") work but couldn't figure it out.  Maybe with your greater familiarity of the project you can find a better solution than the explicit message override I used.

LuRsT replied:
As @gnarea said, we shouldn't have done this in the first place, we are going to revert the specific commit that introduced this problem so we don't need to merge this commit.

Could you remove this commit from your PR? Sorry about the confusion.

jtprince replied:
Yes, will remove.

gnarea commented:
Hi, John!

Thank you so much for your contribution. It looks very good generally, with some minor points that Gil and I raised above. We should be able to get a release out as soon as they're resolved.

Cheers.

jtprince committed:
Deleted .gitignore and updated officially supported versions

Exactly as suggested in both instances.

jtprince commented:
I reverted the change on the HubspotException (removed the 'message' addition).  So, all tests should pass now on python 2.7 (...pending).    However, this still leaves us with one failing test on python 3.4, (as shown below).  I'm hoping you can resolve that one on your end.  In addition, someone might want to update Travis CI to run tests on python 3.4 (I don't think I can do that).

Thanks again!

```
ERROR: tests.test_connection.TestErrorResponses.test_server_error_response
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jtprince/virtualenvs/python3.4-playground/lib/python3.4/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/jtprince/dev/hubspot-connection/tests/test_connection.py", line 244, in test_server_error_response
    eq_('500 Reason', repr(exception))
  File "/home/jtprince/dev/hubspot-connection/hubspot/connection/exc.py", line 67, in __repr__
    return '{} {}'.format(self.http_status_code, self.message)
nose.proxy.AttributeError: 'HubspotServerError' object has no attribute 'message'
-------------------- >> begin captured logging << --------------------
requests.packages.urllib3.util.retry: DEBUG: Converted retries value: 3 -> Retry(total=3, connect=None, read=None, redirect=None)
--------------------- >> end captured logging << ---------------------
```

jtprince commented:
Oh, I should mention that I rebased the history to remove that commit, so _if you were using the DobaTech master_ you will **definitely** want to:

```
git checkout master; git fetch; git reset --hard origin/master
```

jtprince committed:
remove str() invocation for pyrecord arguments

gnarea commented:
Hi John. Thanks for this.

I've created a new branch (`python3`) to get Travis to test this project under Python 3. We should be able to see the results of your changes on this version of Python, but it might require some action on your part.

I think at this point we're waiting for us to deal with a couple of things (the exception and PyRecord), so I've created a separate ticket in our internal bug tracking system to look into those. We should be getting this done by Monday, if not tomorrow.

Cheers.

jtprince commented:
That sounds wonderful.  I really appreciate it.

LuRsT commented on hubspot/connection/__init__.py:
```diff
@@ -13,16 +13,18 @@
 # INFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE.
 #
 ##############################################################################
+from __future__ import unicode_literals
```
This line is the one that is causing problems with the tests. Could you please remove it?

LuRsT replied:
Ignore that, I'll do it :)

LuRsT commented:
Hi @jtprince, thanks again for your contribution, I merged everything into master and will be releasing this shortly :smiley:

jtprince commented:
Thank you all for your help in getting this integrated.  I really appreciate it.  I've also updated hubspot-contacts (was waiting for this one to go through though to use it as a dependency).  I'll be submitting that pull request next week.  Thanks again!

LuRsT commented:
All ready and published successfully. Looking forward to your next Pull Request :+1:
